### ISSUE TITLE
RethinkDB ESM performance

### ISSUE THREAD

**Issue by magaton** [grahamjenson/hapiger#6]
Hello, I am running RethinkDB on my Mac with rethinkdash driver installed and I am evaluating Hapiger performance with rethinkdb as esm.
Write throughput is about 150 events per second. I have 4 different actions.
While writing it is almost impossible to execute recommendation requests - times out after 3min.
When I stop writing with 220K events insterted, recommendation request completes in 30-40s which is very slow. I am looking for <100ms response time.
Anything I can change improve in default config?
Thank you.

**Comment by grahamjenson** [comment]
Hey,
After a recent refactor of the API between GER and the ESMs, the performance of the RethinkDB ESM had a significant impact, and only the PostgresESM is really performant and production ready. 

If you want to contribute any performance enhancements to the RethinkDB ESM it is available at https://github.com/grahamjenson/ger_rethinkdb_esm.

**Comment by magaton** [comment]
Hi, thanks.
I've tried the PG ESM on my local machine (MacBook Pro, 8 GB RAM, 4 dual cores i7, 2.4GHz) and with write throughput of 100 events per second and database with 1M events, it takes 4-5 seconds to get recommendations back. Is this in line with your benchmarks?

**Comment by grahamjenson** [comment]
Hey,

What configuration variables are you using and what domain are the recommendations. By tuning the configurations like neighbourhood_search_size will reduce the time it takes to generate recommendations.

The configurations are listed https://github.com/grahamjenson/ger

Personally I have had good enough recommendations from 250M events in about 300ms, with a lot of tuning.

**Comment by magaton** [comment]
Hi,
retail domain (your sample config):
{
    "namespace": "ACME",
    "person": "rick",
    "configuration": {
      "actions" : {"VIEWED": "1", "LIKED": "2", "ADDED_TO_FAVORITES": "5", "BOUGHT": "10"},
      "minimum_history_required": 5,
  "similarity_search_size": 50,
  "neighbourhood_size": 20,
  "recommendations_per_neighbour": 10,
  "event_decay_rate": 1.05,
  "time_until_expiry": 180
    }
}

**Comment by grahamjenson** [comment]
I would reduce neighbourhood_search_size until you find a speed fast enough. I think the default is 100, that means that it will look in the persons history 100 events to find neighbours, so that is a pretty big join. If you have only a small number of items, like a few hundred, then this join will match just about every person in the dataset. 

Also, if enough of your users actually like, add to favourites, or buy items, then I would drop VIEW as an action. If there are two groups of users, some that interact heavily and some that only view (this is common) then I would try find recommendations with VIEW is excluded first, and if that returns no recommendations (it will return quickly) then do another query that only includes view. 

I cannot give you recommendations for your particular domain, but if you have some data about the dataset I could help a bit more :)

**Comment by magaton** [comment]
Thanks.
I've tried different "neighbourhood_search_size"   and only when it is <20 it gives <500ms response time. Also, I've dropped views, as you suggested, so I am using only 2 event types.
Although idea looks very attractive, I am not so sure postgres esm can scale to millions of events, and 100ms response is my hard SLA. Did you maybe consider using graphdb to get rid of expensive joins?

**Comment by grahamjenson** [comment]
I have not tried a graphdb yet, I looked at Neo4J but I do not think that the types of queries will be speed up by it. You are welcome to try implement your own ESM using such.

Getting good recommendations at under 100ms with millions of records means you will need an in memory solution, not my in memory solution which is more there for test purposes. I recommend looking into implementing an ESM in something like a Redis or memcached, or just improving my in memory implementation.

Another way, would be if your recommendations don't change frequently is to have a big cache, not currently implemented in HapiGER, but something that I would definitly accept a PR for. Hapi has a really easy to use caching which can mean only the first query for recommendations is slow.

100ms is very fast, like I said I got to 300ms with many millions of records before the quality of recommendations were really hurt by the configuration.
